[sw/silicon_creator] Initialize cpuctrl CSR using OTP

alphan opened the pull request:
This PR adds the `CREATOR_SW_CFG_CPUCTRL` OTP item and writes its value to the `cpuctrl` CSR (after masking with 0x3f to preserve the `double_fault_seen` `sync_exc_seen` and bits). This way we can control and test `dummy_instr_mask`, `dummy_instr_en`, and `data_ind_timing` in addition to `icache_enable`.

Please let me know if I should update the mask to preserve the reset values of the bits other than `icache_enable`.

Please see [this page](https://ibex-core.readthedocs.io/en/latest/03_reference/cs_registers.html#cpu-control-register-cpuctrl) for more information on the `cpuctrl` CSR.

This PR also sets the value of this entry to `0x1` in the RMA OTP image. I'll give the resulting bitstream a try before merging this PR.

moidx commented on sw/device/silicon_creator/mask_rom/mask_rom.c:
```diff
@@ -89,6 +89,8 @@ static inline rom_error_t mask_rom_irq_error(void) {
 static rom_error_t mask_rom_init(void) {
   CFI_FUNC_COUNTER_INCREMENT(rom_counters, kCfiRomInit, 1);
+  // Write OTP value to bits 0 to 5 of the cpuctrl CSR.
+  CSR_WRITE(CSR_REG_CPUCTRL, otp_read32(OTP_CTRL_PARAM_CREATOR_SW_CFG_CPUCTRL_OFFSET) & 0x3f);
```
Since this enables some of the Ibex countermeasures, can you also add a couple CSR_READ checks later on to ensure the CSR has the expected value?

alphan replied:
Done

alphan commented:
Thanks @moidx! Rebased, refactored pre-boot checks to `mask_rom_pre_boot_check()`, added additional checks for the cpuctrl CSR.

Relevant part from the [disassembly](https://gist.github.com/78b38bbc7982012810fb84e4062fd714):
```
/proc/self/cwd/sw/device/silicon_creator/mask_rom/mask_rom.c:262
      otp_read32(OTP_CTRL_PARAM_CREATOR_SW_CFG_CPUCTRL_OFFSET);
    8bec:       10c00513                li      a0,268
    8bf0:       165030ef                jal     ra,c554 <otp_read32>
/proc/self/cwd/sw/device/silicon_creator/mask_rom/mask_rom.c:263
  CSR_READ(CSR_REG_CPUCTRL, &cpuctrl_csr);
    8bf4:       7c0025f3                csrr    a1,0x7c0
launder32():
/proc/self/cwd/./sw/device/lib/base/hardened.h:243
    8bf8:       862e                    mv      a2,a1
mask_rom_pre_boot_check():
/proc/self/cwd/sw/device/silicon_creator/mask_rom/mask_rom.c:264
  if (launder32(cpuctrl_csr) != cpuctrl_otp) {
    8bfa:       0ca61c63                bne     a2,a0,8cd2 <mask_rom_boot+0x316>
/proc/self/cwd/sw/device/silicon_creator/mask_rom/mask_rom.c:267
  HARDENED_CHECK_EQ(cpuctrl_csr, cpuctrl_otp);
    8bfe:       00a58663                beq     a1,a0,8c0a <mask_rom_boot+0x24e>
    8c02:       0000                    unimp
    8c04:       0000                    unimp
    8c06:       0000                    unimp
    8c08:       0000                    unimp
```

BTW, the previous initial values for CFI counters were not all 11-bits (at least one of them was 12-bits) contradicting with the comment above the macro.

Also, I'm not sure why there was a inconsistency in formatting. I'll drop the last commit if this was due to a clang-format version mismatch.

alphan commented:
> Also, I'm not sure why there was a inconsistency in formatting. I'll drop the last commit if this was due to a clang-format version mismatch.

I think I know why: the X-macro that defines the initial values for the CFI counters ends with a `\` in `master` eating the `// clang-format on` in the next line.

alphan commented:
I tried the CW310 bitstream with the mask ROM and `CREATOR_SW_CFG_CPUCTRL` OTP item set to `0x1` but I didn't receive anything over UART. Commenting out the `CSR_WRITE()` (or writing `0`) resolves the issue. I reverted the changes in DEV and RMA OTP hjson files, i.e. `CREATOR_SW_CFG_CPUCTRL` is now `0`. I'll try the new bitstream to make sure that it works as expected.

tjaychen commented:
@alphan do you want to push the changes that failed into this PR?
Most likely the failure will be seen in simulation as well, as we can debug it (unless you were seeing the failure only on fpga?)

alphan commented:
> @alphan do you want to push the changes that failed into this PR? Most likely the failure will be seen in simulation as well, as we can debug it (unless you were seeing the failure only on fpga?)

@tjaychen this was most likely a false alarm (reading from the OTP before `sec_mmio_init()`). I'll push an update but things seem to be OK on CW310 locally.

moidx commented on sw/device/silicon_creator/mask_rom/mask_rom.c:
```diff
@@ -283,11 +338,13 @@ static rom_error_t mask_rom_boot(const manifest_t *manifest,
-  sec_mmio_check_values(rnd_uint32());
-  sec_mmio_check_counters(/*expected_check_count=*/4);
+  CFI_FUNC_COUNTER_PREPCALL(rom_counters, kCfiRomBoot, 2, kCfiRomPreBootCheck);
+  mask_rom_pre_boot_check();
```
One last nit: It would be preferable before `flash_ctrl_exec_set` .

alphan replied:
Good point, done.

alphan commented:
```
bazel run //sw/host/opentitantool -- bootstrap --protocol eeprom `pwd`/bazel-bin/sw/device/silicon_creator/rom_ext/virtual_addr_fpga_cw310.test_key_0.signed.bin
```

```
MaskROM:5b2c4370
MaskROM:5b2c4370
starting rom_ext
BFV:0142500d
LCV:2739ce73
```

tjaychen commented:
thanks @alphan i'll update and take a look at the waves.
I'm a little surprised the flash didn't fall over...

tjaychen commented:
o wait, i realized just now we did not apply this change to `test_rom`.
Should we do the same there? Otherwise most of our `DV / fpga` won't get any coverage.

alphan commented:
> @timothytrippel o wait, i realized just now we did not apply this change to `test_rom`. Should we do the same there? Otherwise most of our `DV / fpga` won't get any coverage.

I think it's a good idea, but really up to you. @timothytrippel you can take a look at this change but I'm not sure you'd care about the OTP stuff. Basically all you need is to write 1 to the cpuctrl CSR.

alphan commented:
> thanks @alphan i'll update and take a look at the waves. I'm a little surprised the flash didn't fall over...

We don't have a ton of coverage for this yet but works on my machine™.

tjaychen commented:
@timothytrippel if you're too busy to look at this, i can have a look later today as well.

timothytrippel commented:
Yeah I think it's a good idea to write the cpuctrl register in test ROM too. if we wanted to do an OTP copy tho we would have to update the OTP image that used in testing to include the relevant bits. @alphan when you say we just need to "write 1" to the cpuctrl register, are you saying we would just enable the icache in test ROM? and not configure any of the other bits?

timothytrippel commented:
The benefit of making it the same OTP copy as mask ROM is we get closer feature parity, but we need a better mechanism for generating various OTP images for tests (which is on the ever growing TODO list hehe)

timothytrippel commented:
Yeah I agree, especially with the modular source code bit if we want feature parity. So for the time being maybe we just turn on icache in test ROM and call it a day?

alphan commented:
> Yeah I agree, especially with the modular source code bit if we want feature parity. So for the time being maybe we just turn on icache in test ROM and call it a day?

Sure. We can revisit this discussion when we start touching other bits. @tjaychen WDYT?

timothytrippel commented:
we could make it a configuration parameter like the [`kJitterEnabled`](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/test_rom/test_rom.c;drc=adca27cb824c1c75dd1eec9bf996c4f020f149a1;l=94), and let the test bench backdoor overwrite the symbol. but this will only work on DV platform obviously. not FPGA.

tjaychen commented:
yeah that sounds good to me!
